File: tools/assembler/packages/architecture/riscv/pseudo.py
# -*- coding: utf-8 -*-
# vim: filetype=python
#
# This source file is subject to the MIT License
# that is bundled with this package in the file LICENSE.txt.
# It is also available through the Internet at this address:
# https://opensource.org/license/mit
#
# @author	Sebastien LEGRAND
# @license	MIT License
#
# @brief	Pseudo-Instruction expansion

#----- imports
from typing import List, cast

from packages import ast
from packages.structs import CPU

#----- functions
def instr_expand(cpu: CPU, instr: ast.Instruction) -> List[ast.Instruction]:
    """Expland the pseudo-instruction into standard Risc-V instructions

    Args:
        instr (ast.Instruction): the instruction to expand

    Returns:
        List[ast.Instruction]: the list of instruction following the expansion
    """
    instructions: List[ast.Instruction] = []

    #--- lambdas definition
    # create the label identifier from an operand
    label = lambda x: ast.Identifier(cast(ast.Identifier, instr.operands[x]).name)

    # create the register identifier
    x = lambda n: ast.Identifier(f"x{n}")

    # definition of the number '0'
    zero = ast.Number(0)

    # groups of aliases to expand
    instr_0arg = ['nop', 'ret']
    instr_1arg = ['call', 'j', 'jr']
    instr_2arg = ['mv', 'seqz', 'snez', 'sltz', 'sgtz', 'not', 'neg']
    branch_1 = ['beqz', 'bnez', 'blez', 'bgez', 'bltz', 'bgtz']
    branch_2 = ['bgt', 'ble', 'bgtu', 'bleu']

    if instr.opcode == "li":
        rd, imm = instr.operands

        if isinstance(imm, ast.Number):
            if -2048 <= imm.value <= 2047:
                return [
                    ast.Instruction("addi", [rd, x(0), imm])
                ]

            # big constant
            upper = (imm.value + 0x800) >> 12
            lower = imm.value - (upper << 12)

            return [
                ast.Instruction("lui", [rd, ast.Number(upper)]),
                ast.Instruction("addi", [rd, rd, ast.Number(lower)])
            ]

        if isinstance(imm, ast.Expression):
            return [
                ast.Instruction("lui", [rd, ast.HiRel(imm)]),
                ast.Instruction("addi", [rd, rd, ast.LoRel(imm)])
            ]

    elif instr.opcode == "la":
        rd = instr.operands[0]
        offset = label(1)
        return [
            ast.Instruction('auipc', [rd, ast.PCRelHi(offset)]),
            ast.Instruction('addi', [rd, rd, ast.PCRelLo(offset)])
        ]

    elif instr.opcode.startswith("csr"):
        match instr.opcode:
            case "csrr":
                rd, csr = instr.operands
                return [ ast.Instruction('csrrs', [rd, csr, x(0)])]
            case "csrw":
                csr, rs1 = instr.operands
                return [ ast.Instruction('csrrw', [x(0), csr, rs1])]
            case "csrs":
                csr, rs1 = instr.operands
                return [ ast.Instruction('csrrs', [x(0), csr, rs1])]
            case "csrc":
                csr, rs1 = instr.operands
                return [ ast.Instruction('csrrc', [x(0), csr, rs1])]
            case "csrwi":
                csr, imm = instr.operands
                return [ ast.Instruction('csrrwi', [x(0), csr, imm])]
            case "csrsi":
                csr, imm = instr.operands
                return [ ast.Instruction('csrrsi', [x(0), csr, imm])]
            case "csrci":
                csr, imm = instr.operands
                return [ ast.Instruction('csrrci', [x(0), csr, imm])]

    elif instr.opcode in instr_0arg:
        match instr.opcode:
            case "nop":
                return [ ast.Instruction('addi', [x(0), x(0), zero]) ]
            case "ret":
                return [ast.Instruction("jalr", [ x(0), x(1), zero])]

    elif instr.opcode in instr_1arg:
        offset = label(0)
        match instr.opcode:
            case "call":
                return [
                    ast.Instruction('auipc', [x(1), ast.PCRelHi(offset)]),
                    ast.Instruction('jalr', [x(1), ast.PCRelLo(offset), x(1)])
                ]
            case "j":
                return [ ast.Instruction('jal', [x(0), offset]) ]
            case "jr":
                return [ ast.Instruction('jal', [x(1), offset])]

    elif instr.opcode in instr_2arg:
        rd = instr.operands[0]
        rs = instr.operands[1]

        match instr.opcode:
            case "mv":
                return [ast.Instruction('addi', [rd, rs, zero])]
            case "seqz":
                return [ast.Instruction('sltiu', [rd, rs, ast.Number(1)])]
            case "snez":
                return [ast.Instruction('sltu', [rd, x(0), rs])]
            case "sltz":
                return [ast.Instruction('slt', [rd, rs, x(0)])]
            case "sgtz":
                return [ast.Instruction('slt', [rd, x(0), rs])]
            case "not":
                return [ast.Instruction('xori', [rd, rs, ast.Number(-1)])]
            case "neg":
                return [ast.Instruction('sub', [rd, x(0), rs])]

    elif instr.opcode in branch_1:
        rs = instr.operands[0]
        offset = label(1)

        match instr.opcode:
            case 'beqz':
                return [ast.Instruction('beq', [rs, x(0), offset])]
            case 'bnez':
                return [ast.Instruction('bne', [rs, x(0), offset])]
            case 'blez':
                return [ast.Instruction('bge', [x(0), rs, offset])]
            case 'bgez':
                return [ast.Instruction('bge', [rs, x(0), offset])]
            case 'bltz':
                return [ast.Instruction('blt', [rs, x(0), offset])]
            case 'bgtz':
                return [ast.Instruction('blt', [x(0), rs, offset])]

    elif instr.opcode in branch_2:
        rs = instr.operands[0]
        rt = instr.operands[1]
        offset = label(2)

        match instr.opcode:
            case 'bgt':
                return [ast.Instruction('blt', [rt, rs, offset])]
            case 'ble':
                return [ast.Instruction('bge', [rt, rs, offset])]
            case 'bgtu':
                return [ast.Instruction('bltu', [rt, rs, offset])]
            case 'bleu':
                return [ast.Instruction('bgeu', [rt, rs, offset])]

    # misc pseudo-instruction
    elif instr.opcode == "syscall":
        param = instr.operands[0]
        if isinstance(param, ast.Number):
            return [
                ast.Instruction("addi", [x(17), x(0), param]),
                ast.Instruction("ecall", [])
            ]
        elif isinstance(param, ast.Identifier):
            if (param.name in cpu.registers):
                return [
                    ast.Instruction("add", [x(17), x(0), param]),
                    ast.Instruction("ecall", [])
                ]
            else:
                raise SyntaxError(f"Error: {param.name} is not a valid register in syscall")

    else:
        instructions.append(instr)

    return instructions

File: tools/goasm/internal/decoder/pseudo.go
// Pseudo-instruction recognition for the NATOGA32 disassembler
//
// This source file is subject to the MIT License
// that is bundled with this package in the file LICENSE.txt.
package decoder

import (
	"fmt"

	"github.com/natoga32/goasm/internal/arch"
)

// PseudoMatcher recognizes instruction patterns that map to pseudo-instructions
type PseudoMatcher struct{}

// NewPseudoMatcher creates a new PseudoMatcher
func NewPseudoMatcher() *PseudoMatcher {
	return &PseudoMatcher{}
}

// Match attempts to match a decoded instruction to a pseudo-instruction.
// Returns the pseudo-instruction mnemonic and formatted operands, or empty strings if no match.
func (pm *PseudoMatcher) Match(instr *DecodedInstruction, addr uint32, useABI bool) (mnemonic string, operands string, matched bool) {
	if instr.IsUnknown() {
		return "", "", false
	}

	switch instr.Mnemonic {
	case "addi":
		return pm.matchAddi(instr, useABI)
	case "jalr":
		return pm.matchJalr(instr, useABI)
	case "jal":
		return pm.matchJal(instr, addr, useABI)
	case "beq":
		return pm.matchBeq(instr, addr, useABI)
	case "bne":
		return pm.matchBne(instr, addr, useABI)
	case "blt":
		return pm.matchBlt(instr, addr, useABI)
	case "bge":
		return pm.matchBge(instr, addr, useABI)
	case "bltu":
		return pm.matchBltu(instr, addr, useABI)
	case "bgeu":
		return pm.matchBgeu(instr, addr, useABI)
	case "slt":
		return pm.matchSlt(instr, useABI)
	case "sltu":
		return pm.matchSltu(instr, useABI)
	case "sltiu":
		return pm.matchSltiu(instr, useABI)
	case "sub":
		return pm.matchSub(instr, useABI)
	case "xori":
		return pm.matchXori(instr, useABI)
	}

	return "", "", false
}

// matchAddi handles addi-based pseudo-instructions:
// - nop: addi x0, x0, 0
// - li: addi rd, x0, imm
// - mv: addi rd, rs1, 0
func (pm *PseudoMatcher) matchAddi(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	// nop: addi x0, x0, 0
	if instr.Rd == 0 && instr.Rs1 == 0 && instr.Immediate == 0 {
		return "nop", "", true
	}

	// li: addi rd, x0, imm
	if instr.Rs1 == 0 {
		rd := regName(instr.Rd, useABI)
		return "li", fmt.Sprintf("%s, %d", rd, instr.Immediate), true
	}

	// mv: addi rd, rs1, 0
	if instr.Immediate == 0 {
		rd := regName(instr.Rd, useABI)
		rs1 := regName(instr.Rs1, useABI)
		return "mv", fmt.Sprintf("%s, %s", rd, rs1), true
	}

	return "", "", false
}

// matchJalr handles jalr-based pseudo-instructions:
// - ret: jalr x0, ra, 0
// - jr: jalr x0, rs1, 0
func (pm *PseudoMatcher) matchJalr(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	// Only match when rd=x0 and imm=0
	if instr.Rd != 0 || instr.Immediate != 0 {
		return "", "", false
	}

	// ret: jalr x0, ra, 0
	if instr.Rs1 == 1 { // ra = x1
		return "ret", "", true
	}

	// jr: jalr x0, rs1, 0
	rs1 := regName(instr.Rs1, useABI)
	return "jr", rs1, true
}

// matchJal handles jal-based pseudo-instructions:
// - j: jal x0, offset
// - call: jal ra, offset
func (pm *PseudoMatcher) matchJal(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	target := calculateTarget(addr, instr.Immediate)

	// j: jal x0, offset
	if instr.Rd == 0 {
		return "j", fmt.Sprintf("0x%X", target), true
	}

	// call: jal ra, offset
	if instr.Rd == 1 { // ra = x1
		return "call", fmt.Sprintf("0x%X", target), true
	}

	return "", "", false
}

// matchBeq handles beq-based pseudo-instructions:
// - beqz: beq rs1, x0, offset
func (pm *PseudoMatcher) matchBeq(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	// beqz: beq rs1, x0, offset
	if instr.Rs2 == 0 {
		target := calculateTarget(addr, instr.Immediate)
		rs1 := regName(instr.Rs1, useABI)
		return "beqz", fmt.Sprintf("%s, 0x%X", rs1, target), true
	}

	return "", "", false
}

// matchBne handles bne-based pseudo-instructions:
// - bnez: bne rs1, x0, offset
func (pm *PseudoMatcher) matchBne(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	// bnez: bne rs1, x0, offset
	if instr.Rs2 == 0 {
		target := calculateTarget(addr, instr.Immediate)
		rs1 := regName(instr.Rs1, useABI)
		return "bnez", fmt.Sprintf("%s, 0x%X", rs1, target), true
	}

	return "", "", false
}

// matchBlt handles blt-based pseudo-instructions:
// - bgtz: blt x0, rs2, offset
// - bltz: blt rs1, x0, offset
// - bgt: blt rs2, rs1, offset (swapped operands)
func (pm *PseudoMatcher) matchBlt(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	target := calculateTarget(addr, instr.Immediate)

	// bgtz: blt x0, rs2, offset
	if instr.Rs1 == 0 && instr.Rs2 != 0 {
		rs2 := regName(instr.Rs2, useABI)
		return "bgtz", fmt.Sprintf("%s, 0x%X", rs2, target), true
	}

	// bltz: blt rs1, x0, offset
	if instr.Rs2 == 0 && instr.Rs1 != 0 {
		rs1 := regName(instr.Rs1, useABI)
		return "bltz", fmt.Sprintf("%s, 0x%X", rs1, target), true
	}

	// bgt: blt rs2, rs1, offset (swapped operands, both non-zero)
	if instr.Rs1 != 0 && instr.Rs2 != 0 {
		// Swap: blt rs1, rs2 => bgt rs2, rs1
		rs1 := regName(instr.Rs1, useABI)
		rs2 := regName(instr.Rs2, useABI)
		return "bgt", fmt.Sprintf("%s, %s, 0x%X", rs2, rs1, target), true
	}

	return "", "", false
}

// matchBge handles bge-based pseudo-instructions:
// - bgez: bge rs1, x0, offset
// - blez: bge x0, rs2, offset
// - ble: bge rs2, rs1, offset (swapped operands)
func (pm *PseudoMatcher) matchBge(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	target := calculateTarget(addr, instr.Immediate)

	// bgez: bge rs1, x0, offset
	if instr.Rs2 == 0 && instr.Rs1 != 0 {
		rs1 := regName(instr.Rs1, useABI)
		return "bgez", fmt.Sprintf("%s, 0x%X", rs1, target), true
	}

	// blez: bge x0, rs2, offset
	if instr.Rs1 == 0 && instr.Rs2 != 0 {
		rs2 := regName(instr.Rs2, useABI)
		return "blez", fmt.Sprintf("%s, 0x%X", rs2, target), true
	}

	// ble: bge rs2, rs1, offset (swapped operands, both non-zero)
	if instr.Rs1 != 0 && instr.Rs2 != 0 {
		// Swap: bge rs1, rs2 => ble rs2, rs1
		rs1 := regName(instr.Rs1, useABI)
		rs2 := regName(instr.Rs2, useABI)
		return "ble", fmt.Sprintf("%s, %s, 0x%X", rs2, rs1, target), true
	}

	return "", "", false
}

// matchBltu handles bltu-based pseudo-instructions:
// - bgtu: bltu rs2, rs1, offset (swapped operands)
func (pm *PseudoMatcher) matchBltu(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	// bgtu: bltu rs2, rs1, offset (swapped operands, both non-zero)
	if instr.Rs1 != 0 && instr.Rs2 != 0 {
		target := calculateTarget(addr, instr.Immediate)
		rs1 := regName(instr.Rs1, useABI)
		rs2 := regName(instr.Rs2, useABI)
		return "bgtu", fmt.Sprintf("%s, %s, 0x%X", rs2, rs1, target), true
	}

	return "", "", false
}

// matchBgeu handles bgeu-based pseudo-instructions:
// - bleu: bgeu rs2, rs1, offset (swapped operands)
func (pm *PseudoMatcher) matchBgeu(instr *DecodedInstruction, addr uint32, useABI bool) (string, string, bool) {
	// bleu: bgeu rs2, rs1, offset (swapped operands, both non-zero)
	if instr.Rs1 != 0 && instr.Rs2 != 0 {
		target := calculateTarget(addr, instr.Immediate)
		rs1 := regName(instr.Rs1, useABI)
		rs2 := regName(instr.Rs2, useABI)
		return "bleu", fmt.Sprintf("%s, %s, 0x%X", rs2, rs1, target), true
	}

	return "", "", false
}

// matchSlt handles slt-based pseudo-instructions:
// - sltz: slt rd, rs1, x0
// - sgtz: slt rd, x0, rs1
func (pm *PseudoMatcher) matchSlt(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	rd := regName(instr.Rd, useABI)

	// sltz: slt rd, rs1, x0
	if instr.Rs2 == 0 && instr.Rs1 != 0 {
		rs1 := regName(instr.Rs1, useABI)
		return "sltz", fmt.Sprintf("%s, %s", rd, rs1), true
	}

	// sgtz: slt rd, x0, rs1
	if instr.Rs1 == 0 && instr.Rs2 != 0 {
		rs2 := regName(instr.Rs2, useABI)
		return "sgtz", fmt.Sprintf("%s, %s", rd, rs2), true
	}

	return "", "", false
}

// matchSltu handles sltu-based pseudo-instructions:
// - snez: sltu rd, x0, rs1
func (pm *PseudoMatcher) matchSltu(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	// snez: sltu rd, x0, rs1
	if instr.Rs1 == 0 && instr.Rs2 != 0 {
		rd := regName(instr.Rd, useABI)
		rs2 := regName(instr.Rs2, useABI)
		return "snez", fmt.Sprintf("%s, %s", rd, rs2), true
	}

	return "", "", false
}

// matchSltiu handles sltiu-based pseudo-instructions:
// - seqz: sltiu rd, rs1, 1
func (pm *PseudoMatcher) matchSltiu(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	// seqz: sltiu rd, rs1, 1
	if instr.Immediate == 1 {
		rd := regName(instr.Rd, useABI)
		rs1 := regName(instr.Rs1, useABI)
		return "seqz", fmt.Sprintf("%s, %s", rd, rs1), true
	}

	return "", "", false
}

// matchSub handles sub-based pseudo-instructions:
// - neg: sub rd, x0, rs1
func (pm *PseudoMatcher) matchSub(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	// neg: sub rd, x0, rs1
	if instr.Rs1 == 0 && instr.Rs2 != 0 {
		rd := regName(instr.Rd, useABI)
		rs2 := regName(instr.Rs2, useABI)
		return "neg", fmt.Sprintf("%s, %s", rd, rs2), true
	}

	return "", "", false
}

// matchXori handles xori-based pseudo-instructions:
// - not: xori rd, rs1, -1
func (pm *PseudoMatcher) matchXori(instr *DecodedInstruction, useABI bool) (string, string, bool) {
	// not: xori rd, rs1, -1
	if instr.Immediate == -1 {
		rd := regName(instr.Rd, useABI)
		rs1 := regName(instr.Rs1, useABI)
		return "not", fmt.Sprintf("%s, %s", rd, rs1), true
	}

	return "", "", false
}

// regName returns the register name based on the useABI flag
func regName(reg int, useABI bool) string {
	return arch.RegisterName(reg, useABI)
}

// calculateTarget calculates the absolute target address from PC and offset
func calculateTarget(addr uint32, offset int32) uint32 {
	return uint32(int32(addr) + offset)
}
